File: my-folder/0093-restore-ip-addresses/solution.go
func restoreIpAddresses(s string) []string {
    res := []string{}
    if len(s) > 12 {
        return res
    }

    var dfs func(i, dots int, currIP string)
    dfs = func(i, dots int, currIP string) {
        if dots == 4 && i == len(s) {
            res = append(res, currIP[:len(currIP)-1])
            return
        }
        if dots > 4 {
            return
        }

        for j := i; j < min(i+3, len(s)); j++ {
            if i != j && s[i] == '0' {
                continue
            }
            part := s[i : j+1]
            num, _ := strconv.Atoi(part)
            if num < 256 {
                dfs(j+1, dots+1, currIP+part+".")
            }
        }
    }

    dfs(0, 0, "")
    return res
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

File: my-folder/0093-restore-ip-addresses/solution.ts
function restoreIpAddresses(s: string): string[] {
    let res: string[] = []

    if (s.length > 12) {
        return res
    }

    let backtracking = function(i: number, dots: number, currIP: string) {
        if (dots === 4 && i === s.length) {
            res.push(currIP.slice(0, currIP.length-1))
            return
        }

        if (dots > 4)
            return

        for (let j = i; j < Math.min(i+3, s.length); j++) {
            if (i !== j && s[i] === "0") {
                continue
            }
            let num = parseInt(s.slice(i, j+1))
            if (num < 256) {
                backtracking(j+1, dots+1, currIP+num+".")
            }
        }
    }

    backtracking(0, 0, "")
    return res
};
